langchain4j/src/test/java/dev/langchain4j/service/AiServicesWithToolsIT.java

```diff
@@ -681,6 +673,126 @@ void should_use_programmatically_configured_tools(ChatLanguageModel chatLanguage
+        ToolSpecification toolSpecification = ToolSpecification.builder()
+                .name("get_booking_details")
+                .description("Returns booking details")
+                .addParameter("bookingNumber", type("string"))
+                .build();
+
+        tools.add(toolSpecification, executor);
+
+        Assistant assistant = AiServices.builder(Assistant.class)
+                .chatLanguageModel(chatLanguageModel)
+                .build();
+
+        // Tool shouldn't be used
+        assistant.chat("When does my booking 123-456 starts?");
+        verifyNoInteractions(executor);
+
+        // Verify tool has been used
+        Response<AiMessage> response = assistant.chat("When does my booking 123-456 starts?", tools);
+        verify(executor, times(1)).execute(any(), any());
+        assertThat(response.content().text()).contains("2027");
+
+        // Tool shouldn't be used again. -> Normal again.
+        Response<AiMessage> responseNoTools2 = assistant.chat("When does my booking 123-456 starts?");
+        verify(executor, times(1)).execute(any(), any());
+        assertThat(responseNoTools2.content().text()).doesNotContain("2027");
+    }
+
+    @ParameterizedTest
+    @MethodSource("models")
+    void should_use_tool_provider(ChatLanguageModel chatLanguageModel) {
+        ToolProviderResult bookingResult = new ToolProviderResult();
+        ToolExecutor executor = spy(new BookingExecutor());
+        ToolSpecification specification = ToolSpecification.builder()
+                .name("get_booking_details")
+                .description("Returns booking details")
+                .addParameter("bookingNumber", type("string"))
+                .build();
+
+        bookingResult.add(specification, executor);
+
+        Assistant assistant = AiServices.builder(Assistant.class)
+                .chatLanguageModel(chatLanguageModel)
+                .toolsToolProvider((ToolProviderRequest request) -> {
+                    if (request.getUserMessage().singleText().contains("booking")) {
+                        return bookingResult;
+                    }
+                    return null;
+                })
+                .build();
+
+        // The provider will be triggered only on the word booking. Not holiday -> Don't use the tool
+        assistant.chat("When does my holiday 123-456 starts?");
+        verifyNoInteractions(executor);
+
+        // Booking will trigger the tool.
+        Response<AiMessage> response = assistant.chat("When does my booking 123-456 starts?");
+        verify(executor, times(1)).execute(any(), any());
+        assertThat(response.content().text()).contains("2027");
+
+        // No tool again
+        assistant.chat("When does my holiday 123-456 starts?");
+        verify(executor, times(1)).execute(any(), any());
+    }
+
+    @ParameterizedTest
+    @MethodSource("models")
+    void only_static_or_tool_provider(ChatLanguageModel chatLanguageModel) {
```

Comment:
```suggestion
    void only_static_or_tool_provider() {
```
This test does not require testing with various (and real) models, `ChatModelMock` can be used (or `ChatLanguageModel` mocked with Mockito)